File: backend/app/api/runs.py
from typing import Optional

from fastapi import APIRouter, BackgroundTasks, Query, Response

from app.config import get_settings
from app.schemas import RunCreate, RunRead, RunResultRead, ViewerAdjustmentPayload, ViewerAdjustmentResponse
from app.services import run_service

router = APIRouter(prefix="/runs", tags=["runs"])


@router.post("", response_model=RunRead, status_code=201)
def create_run(
    run_create: RunCreate,
    background_tasks: BackgroundTasks,
    response: Response,
):
    out = run_service.create_run(run_create, background_tasks)
    if get_settings().RUN_EXECUTOR == "arq":
        response.status_code = 202
    return out


@router.get("", response_model=list[RunRead])
def list_runs(
    pipeline_id: Optional[str] = Query(default=None),
    homework_id: Optional[str] = Query(default=None),
    status: Optional[str] = Query(default=None),
):
    return run_service.list_runs(pipeline_id=pipeline_id, homework_id=homework_id, status=status)


@router.get("/{run_id}", response_model=RunRead)
def get_run(run_id: str):
    return run_service.get_run(run_id)


@router.get("/{run_id}/results", response_model=list[RunResultRead])
def get_run_results(run_id: str):
    return run_service.get_run_results(run_id)


@router.get("/{run_id}/viewer-adjustments", response_model=ViewerAdjustmentResponse)
def get_viewer_adjustments(run_id: str):
    return run_service.get_viewer_adjustments(run_id)


@router.put("/{run_id}/viewer-adjustments", response_model=ViewerAdjustmentResponse)
def put_viewer_adjustments(run_id: str, body: ViewerAdjustmentPayload):
    return run_service.put_viewer_adjustments(run_id, body)

File: frontend/src/api/runs.ts
import { apiGet, apiPost, apiPut } from './client';
import type { Run, RunResult, CompareResponse, ViewerAdjustmentPayload, ViewerAdjustmentResponse } from '@/types';

interface RunFilters {
  pipeline_id?: string;
  homework_id?: string;
  status?: string;
}

export function listRuns(filters?: RunFilters): Promise<Run[]> {
  const params = new URLSearchParams();
  if (filters?.pipeline_id) params.set('pipeline_id', filters.pipeline_id);
  if (filters?.homework_id) params.set('homework_id', filters.homework_id);
  if (filters?.status) params.set('status', filters.status);
  const query = params.toString();
  return apiGet<Run[]>(`/runs${query ? `?${query}` : ''}`);
}

export function getRun(id: string): Promise<Run> {
  return apiGet<Run>(`/runs/${id}`);
}

export function getRunResults(id: string): Promise<RunResult[]> {
  return apiGet<RunResult[]>(`/runs/${id}/results`);
}

export function createRun(
  pipelineId: string,
  homeworkId: string,
  inferenceProfileId?: string | null,
  pipelineName?: string,
  pipelineVersion?: string,
  pipelineConfig?: Record<string, unknown> | null,
): Promise<Run> {
  return apiPost<Run>('/runs', {
    pipeline_id: pipelineId,
    pipeline_name: pipelineName,
    pipeline_version: pipelineVersion,
    homework_id: homeworkId,
    inference_profile_id: inferenceProfileId ?? undefined,
    pipeline_config: pipelineConfig ?? undefined,
  });
}

export function compareRuns(runIds: string[]): Promise<CompareResponse> {
  const params = new URLSearchParams({ run_ids: runIds.join(',') });
  return apiGet<CompareResponse>(`/compare?${params.toString()}`);
}

export function getViewerAdjustments(runId: string): Promise<ViewerAdjustmentResponse> {
  return apiGet<ViewerAdjustmentResponse>(`/runs/${runId}/viewer-adjustments`);
}

export function putViewerAdjustments(
  runId: string,
  payload: ViewerAdjustmentPayload,
): Promise<ViewerAdjustmentResponse> {
  return apiPut<ViewerAdjustmentResponse>(`/runs/${runId}/viewer-adjustments`, payload);
}
